### ISSUE TITLE
Separate error code for `cargo install` failing due to binary already installed

### ISSUE THREAD

**Issue by djeedai** [rust-lang/cargo#11513]
### Problem

I'm trying to use `cargo install` in CI on GitHub, to install `cargo-tarpaulin`. However `cargo install` might fail if the CI agent already has the version of `cargo-tarpaulin` that I need. Since it takes a bit to install, I also want to use GitHub caching, so want to avoid `--force`. However, I still want to fail CI if the install fails for any other reason than "the version is already installed". Unfortunately `cargo` returns a single error code so there's no way for me to discriminate those two failures, and no easy way that I know of either to check if a binary is installed (`cargo install --list` is hard to parse via script due to its nested listing which outputs both the package and its binaries, which are often the same).

### Proposed Solution

Have `cargo` return a different error code when not using `--force` and the version of the binary is already installed.

### Notes

Briefly discussed on Discord [here](https://discord.com/channels/442252698964721669/443150878111694848/1057364363876519986).

**Comment by weihanglo** [comment]
Out of curiosity, which version are you using? IIRC from Rust 1.41, `cargo install` automatically upgrade installed binaries by default. It should work as the following:

```console
$ cargo install --list
cargo-expand v1.0.36:
    cargo-expand
$ cargo install cargo-expand
    Updating crates.io index
     Ignored package `cargo-expand v1.0.36` is already installed, use --force to override
$ echo $?
0
```

**Comment by djeedai** [comment]
This is on `latest update on 2022-12-15, rust version 1.66.0 (69f9c33d7 2022-12-12)`.

But I suspect then that there's a bad interaction with the `actions/cache@v2` of the GitHub Action, which probably restores the binary but not some config file where that install is logged, yielding a "genuine" error of "cargo didn't expect that binary to be installed, but found it on disk anyway". Here's the GitHub cache action, which incidentally doesn't cache `.crates.toml` nor `.crates2.json`:

https://github.com/djeedai/bevy_hanabi/blob/1713e0cd951fdf265bc9f987a4801b4593138bef/.github/workflows/ci.yaml#L76-L84

**Comment by ehuss** [comment]
I would recommend updating the cache action to include the `.crates2.json` file in order for Cargo to know whether or not it should try to reinstall. There's more documentation of which files cargo uses in https://doc.rust-lang.org/cargo/guide/cargo-home.html.

**Comment by djeedai** [comment]
@ehuss that documentation explicitly lists the things to cache and neither `.crates.toml` nor `.crates2.json` are in that list. Is the documentation incorrect then?

**Comment by weihanglo** [comment]
It is telling caching things to avoid redownload not rebuild executables. I guess that's the difference. Perhaps we can expand the paragraph with some sentenses about how to cache installed binaries.

**Comment by djeedai** [comment]
> Perhaps we can expand the paragraph with some sentenses about how to cache installed binaries.

I think that would be beneficial yes. I've actually copied my own CI settings from another project so I'm not alone making that mistake. How can we determine the set of files which is necessary to cache for installed binaries? I don't even understand `.crates.toml` and `.crates2.json` and why there's 2 files seemingly doing the same thing just in a different format.

**Comment by weihanglo** [comment]
`.crates2.json` is the second version of the tracking file of installed info, which actually does what you originally requested (see https://github.com/rust-lang/cargo/issues/6797). These two files are in sync and have sufficient information to cache installed binaries I believe (see config value [`install.root`](https://doc.rust-lang.org/cargo/reference/config.html#installroot)).

https://github.com/rust-lang/cargo/blob/7fb01c68c190c28a43581ae7c719dcf057ecc002/src/cargo/ops/common_for_install_and_uninstall.rs#L23-L35

It has been three years since the feature got stabilized. Maybe it's about time to remove v1 tracking file.

**Comment by weihanglo** [comment]
Close as #11592 added a tip to cache those two tracking files. Thank you everyone for participating the discussion :)

**Comment by djeedai** [comment]
Thanks for the fix!

**Comment by djeedai** [comment]
Just to circle back, I've finally updated the CI of my project and confirmed that even without the `--force` on the cargo install the binary is properly cached when the GitHub cache action includes `.crates.toml` and `.crates2.json`, slashing 3 minutes in the CI build. Thanks everyone.
